### ISSUE TITLE
CompletableFuture method is not generated for lists

### ISSUE THREAD

**Issue by DimaSol** [graphql-java-generator/graphql-maven-plugin-project#118]
Hi,
I've noticed that if I have a field which is a list of some type (that has an ID) a CompletableFuture for this list items is not generated.
I'm wondering if there is a reason for this?

**Comment by etienne-sf** [comment]
No there is no (valid) reasons.

I'll dig into this, but it looks very much like a bug

**Comment by shaul-fb** [comment]
Hi, any update with that?

**Comment by mheidt** [comment]
@etienne-sf, please make it configurable if you change anything.
I thought, this was a specification thing and I wrote my custom wiring for those cases I wanted to have batchloading on a 1:n relationship.
If you change it automatically, I would have to do a major refactoring.
That doesn't mean, that I would not like the feature. I never understood, why those cases haven't had a CompletableFuture.

I just can say, that it is rather complicated..maybe that's why you didn't do it.
In the end it's n:m and I was using something like this in the core:
`CompletableFuture<Void> allFuturesResult =
 CompletableFuture.allOf(completableFutures.toArray(new CompletableFuture[completableFutures.size()]));
 CompletableFuture<List<T>> completableFutureOfList = allFuturesResult.thenApply(v ->
       completableFutures.stream()
            .map(future -> future.join()).
                      collect(Collectors.<T>toList())
       );
`

**Comment by mheidt** [comment]
Yes, in order to avoid the doublon-issue, I didn't yield for the 100% solution but calling n-calls in parallel instead.
You would need to collect the n id-lists first and then remap...it's on my todo list still but is deferred until I have to make the calls faster :)

**Comment by shaul-fb** [comment]
What I had in mind is to wait for all the ids, and then map them back to the original requests and "duplicate" results if needed and that way avoid N queries to the database. (We do need duplicates).
However, the issue I'm facing now is that I need to write my `completeablefuture` method myself but on every build, the interfaces and the `GraphQLDataFetchers` file are being rebuild by the library and deletes my code.
How do I overcome this?

**Comment by mheidt** [comment]
First rule of fight club: Never touch generated code.
Maybe this helps:
```
@Component
@Primary
public class CustomGraphQLWiring extends GraphQLWiring {

    @Autowired
    CustomGraphQLDataFetchers graphQLDataFetchers;


    @Override
    protected TypeRuntimeWiring.Builder addWiringForApplicationVersionMM(TypeRuntimeWiring.Builder typeWiring) {
        typeWiring = super.addWiringForApplicationVersionMM(typeWiring);
        typeWiring.dataFetcher("applicationVersionCustomerState", graphQLDataFetchers.dataFetchersDelegateApplicationVersionMMApplicationVersionCustomerStatessWithDataLoader());
        return typeWiring;
    }
```

**Comment by shaul-fb** [comment]
@mheidt Thanks for that.
I'm new to Java and Spring. Can you explain what you did and elaborate a bit more?

**Comment by mheidt** [comment]
You need to inherit from the generated classes and mark them as @Primary.

**Comment by DimaSol** [comment]
@etienne-sf 
Pretty important for me.
In my case the gql server is communicating with other service (not DB) and fetching info from it so I'm performing multiple hits to another (external) service that also has some thresholds & throttling  business which means this multiple hits may lead to 429 (Too Many Requests) http code. 
Anyway, I can check if I can overcome as suggested above if a fix is not planned for the near future?

**Comment by etienne-sf** [comment]
Hello,

I reviewed the subject. And I'm not convinced about the gain to have.


Here is a sample, from the forum GraphQL schema sample, used in the project:
```
type Topic {
    id: ID!
    date: Date!
    author: Member!
    publiclyAvailable: Boolean
    nbPosts: Int
    title: String!
    content: String
    posts(memberId: ID, memberName: String, since: Date!): [Post]!
}
```

In its current implementation, a data loader is used for the `author` field, but none for the `posts` field.
This reflects this reality:
* The Topic object contains the id of the author. So adding this id to the data loader, for a further mass retrieval of members is a real optimization.
* The Topic object doesn't know the id of the related posts (whether it's a reational database or REST entity).
    * So, to feed the data loader, you need to execute a first request to get the list of ids (for each topic that ask for a list of posts in the current GraphQL request). Then a second request to read the ids.
    * Whereas, in the current implementation, you'll have only the first request to execute. It will directly retrieve all the post's data.
 

Can you precise your use case (actually, how the data is stored), and why there is an optimization in the number of data accesses ?

Etienne

**Comment by mheidt** [comment]
You are right, when all calls have the same weight.
But in my case, the m 1:n calls to get the IDs are much cheaper than the one, getting the actual result.

**Comment by DimaSol** [comment]
In my case, I do have the ids of the external entities (posts in your example) that were stored previously when user configured and selected some of the external entities in ui (they were retrieved in other query) and in later requests I need to enrich ids with real information upon requested fields.
I believe that a similar use case is relevant for any integration with external systems...

**Comment by etienne-sf** [comment]
> You are right, when all calls have the same weight.
> 
> But in my case, the m 1:n calls to get the IDs are much cheaper than the one, getting the actual result.
> 
>  

Yes. But ... with this asked update, you'll have to execute both queries, instead of just the heavy one. Whether or not there is a win, depends on the number of ids that will get fetched once only, instead of several times. 


BTW, after your two comments, I'll finish the job. 

I'll just change the name of the parameter that activate or deactivate it: its use is not a matter of legacy (the current behavior) or optimized, as I thought. 
Its use rather depends on the way the data can be requested. 

So I guess I'll call the plug-in parameter something like ´generateDataLoaderForLists´ 

Étienne

**Comment by mheidt** [comment]
But shouldn't you be doing it for all but with an annotation for some?

**Comment by etienne-sf** [comment]
I've almost finished the global parameter. 
And yes, a local parameter would be nice. 
I'll see that.

**Comment by etienne-sf** [comment]
Hello

I just pushed these commits:
* Added a `generateDataLoaderForLists` plugin parameter. The default value is false, which let the generated code unchanged. When set to true, data loader methods is also generated for lists (in DataFetcherDelegates)
* Added the `GenerateDataLoaderForLists` directive: it can be added to a specific GraphQL field, which type is a list of GraphQL objects, which have an id. The generated code will then make use of a Data Loader. Please read the details are on the [Wiki server page](https://github.com/graphql-java-generator/graphql-maven-plugin-project/wiki/server).
* The `DataFetcherDelegate` interfaces has now two methods for batch loading (instead of once before). The provided default implementation prevent side effect on the existing `DataFetcherDelegate` implementations. These methods are `batchLoader` and `unorderedReturnBatchLoader`. All the details are on the [Wiki server page](https://github.com/graphql-java-generator/graphql-maven-plugin-project/wiki/server).
* Upgrade of java-dataloader from version 2.2.3 to version 3.1.2

Please note that the wiki server page is not updated yet.

I'll work on issue 127, then I'll release a new version.

**Comment by etienne-sf** [comment]
It's available in the 1.18.4
